feat: Add GnuPG 2.4+ compatibility and fix keychain permission issues

lujstn opened the pull request:
This commit implements critical fixes to make pinentry-touchid compatible with modern GnuPG versions (2.4+) and resolves keychain access permission issues.

## Key Changes

### 1. GETINFO Command Support
- Implemented GETINFO handler for commands required by GnuPG 2.4+
- Added support for: flavor, version, pid, and ttyinfo queries
- Resolves "You may want to update to a newer pinentry" errors

### 2. Keychain Permission Handling
- Added ensureKeychainAccess() function to properly handle macOS permission dialogs
- Fixed flow to check keychain permissions before Touch ID authentication
- Ensures pinentry-touchid can access entries created by pinentry-mac

### 3. Name Parsing Fix
- Fixed regex parsing to correctly extract user names from GPG descriptions
- Strips GPG key comments (e.g., "GPG Key generated...") from names
- Ensures keychain entry labels match between pinentry-mac and pinentry-touchid

### 4. Duplicate Entry Handling
- Changed duplicate keychain entry errors from fatal to non-fatal
- Allows the authentication flow to continue when entries already exist
- Improves robustness when switching between pinentry programs

## Technical Details

The main issue was that GnuPG 2.4+ requires pinentry programs to respond to GETINFO queries during initialization. Without these handlers, gpg-agent would abort the connection immediately after the handshake.

Additionally, when keychain entries are created by pinentry-mac, macOS requires explicit permission for pinentry-touchid to access them. The new flow ensures these permissions are granted before attempting Touch ID authentication.

## Testing

- Added comprehensive unit tests for GETINFO commands
- Verified compatibility with GnuPG 2.4.8 on macOS
- Tested keychain permission flow with existing entries
- All existing tests continue to pass

Fixes #9, #17, #42

lujstn committed:
feat: Add GnuPG 2.4+ compatibility and fix keychain permission issues

This commit implements critical fixes to make pinentry-touchid compatible with modern GnuPG versions (2.4+) and resolves keychain access permission issues.

## Key Changes

### 1. GETINFO Command Support
- Implemented GETINFO handler for commands required by GnuPG 2.4+
- Added support for: flavor, version, pid, and ttyinfo queries
- Resolves "You may want to update to a newer pinentry" errors

### 2. Keychain Permission Handling
- Added ensureKeychainAccess() function to properly handle macOS permission dialogs
- Fixed flow to check keychain permissions before Touch ID authentication
- Ensures pinentry-touchid can access entries created by pinentry-mac

### 3. Name Parsing Fix
- Fixed regex parsing to correctly extract user names from GPG descriptions
- Strips GPG key comments (e.g., "GPG Key generated...") from names
- Ensures keychain entry labels match between pinentry-mac and pinentry-touchid

### 4. Duplicate Entry Handling
- Changed duplicate keychain entry errors from fatal to non-fatal
- Allows the authentication flow to continue when entries already exist
- Improves robustness when switching between pinentry programs

## Technical Details

The main issue was that GnuPG 2.4+ requires pinentry programs to respond to GETINFO queries during initialization. Without these handlers, gpg-agent would abort the connection immediately after the handshake.

Additionally, when keychain entries are created by pinentry-mac, macOS requires explicit permission for pinentry-touchid to access them. The new flow ensures these permissions are granted before attempting Touch ID authentication.

## Testing

- Added comprehensive unit tests for GETINFO commands
- Verified compatibility with GnuPG 2.4.8 on macOS
- Tested keychain permission flow with existing entries
- All existing tests continue to pass

Fixes #9, #17, #42

lujstn committed:
chore: Update module path and add Homebrew tap installation

- Changed module path from jorgelbg to lujstn
- Updated .goreleaser.yml to point to lujstn/homebrew-tap
- Added Homebrew installation instructions to README
- Updated imports to use new module path

lujstn committed:
chore: Update .goreleaser.yml to v2 format

fzipi commented on .goreleaser.yml:
```diff
@@ -20,12 +22,12 @@ builds:
       token: "{{ .Env.HOMEBREW_TAP_GITHUB_TOKEN }}"
 
-    homepage: "https://github.com/jorgelbg/pinentry-touchid"
+    homepage: "https://github.com/lujstn/pinentry-touchid"
```
I don't think this change will fly here, right?

lujstn replied:
Oh yikes - that should only be for my fork! Not trying to be sneaky haha

lujstn replied:
Will push a fix to this branch

lujstn committed:
fix: never exit before Assuan handshake on log file errors

Replace panic() in New() with stderr fallback when the log file cannot be created or opened (e.g. macOS cleaning /private/tmp). Add missing return after Touch ID unavailable fallback to prevent double-serve. Add -self-test flag for startup diagnostics.

fzipi commented:
@jorgelbg can we give this a go?

lujstn committed:
chore: strip non-contributal elements

lujstn commented:
All updated! Also pushed a fix for a couple of bugs I spotted myself in day-to-day usage.

@jorgelbg: What's the status of this repo now? I'm conscious there's been no activity in the past 4yrs so there's 18 issues / 6 PRs unreviewed. I'd be happy to continue actively maintaining and developing this repo if you'd be open to that? Let me know what you think 🙂

fzipi commented:
BTW, I can confirm this fixes GPG. I've switched to lujstn's tap in the meantime.
